src/client_side.cc

```diff
@@ -1638,7 +1638,10 @@ clientProcessRequest(ConnStateData *conn, const Http1::RequestParserPointer &hp,
 #endif
 
     if (internalCheck(request->url.path())) {
-        if (internalHostnameIs(request->url.host()) && request->url.port() == getMyPort()) {
+        const bool allowLocalhost = (conn->port->s.isLocalhost() || conn->port->s.isAnyAddr());
+        const bool isMyHostname = (allowLocalhost && strcmp(request->url.host(), "localhost") == 0) ||
+                                  internalHostnameIs(request->url.host());
+        if (isMyHostname && request->url.port() == conn->port->s.port()) {
             debugs(33, 2, "internal URL found: " << request->url.getScheme() << "://" << request->url.authority(true));
```

Comment:
> We can wait for the full refactor requested.

Does this updated PR represent that "full refactor"? Or should we still treat this PR as a temporary hack to relieve bug 5283 pressures?

Reply:
The parts of the refactor which were colliding with this have been merged as 101694e4332f35dc3b5e4729839fcd2daad858a5 allowing this PR to progress.